core/arch/arm/kernel/user_ta.c

```diff
@@ -233,29 +212,57 @@ static TEE_Result load_elf(struct user_ta_ctx *utc,
 	 * Map physical memory into TA virtual memory
 	 */
 
-	res = tee_mmu_init(utc);
+	res = vm_info_init(utc);
 	if (res != TEE_SUCCESS)
 		goto out;
 
-	res = load_elf_segments(utc, elf_state, true /* init attrs */);
+	/*
+	 * Add stack segment
+	 */
+	utc->stack_addr = 0;
+	res = vm_map(utc, &utc->stack_addr, utc->mobj_stack->size,
+		     TEE_MATTR_URW | TEE_MATTR_PRW, utc->mobj_stack, 0);
+	if (res)
+		goto out;
+
+	res = get_elf_segments(utc, elf_state, &segs, &num_segs);
 	if (res != TEE_SUCCESS)
 		goto out;
 
+	utc->load_addr = 0;
+	for (n = 0; n < num_segs; n++) {
+		uint32_t prot = elf_flags_to_mattr(segs[n].flags) |
+				TEE_MATTR_PRW;
+
+		segs[n].va = utc->load_addr - segs[0].offs + segs[n].offs;
+		segs[n].size = segs[n].oend - segs[n].offs;
+		res = vm_map(utc, &segs[n].va, segs[n].size, prot,
+			     utc->mobj_code, segs[n].offs);
+		if (res)
+			goto out;
+		if (!n)
+			utc->load_addr = segs[0].va;
+	}
+
 	tee_mmu_set_ctx(&utc->ctx);
 
-	res = elf_load_body(elf_state, tee_mmu_get_load_addr(&utc->ctx));
+	res = elf_load_body(elf_state, utc->load_addr);
 	if (res != TEE_SUCCESS)
 		goto out;
 
 	/*
 	 * Replace the init attributes with attributes used when the TA is
 	 * running.
 	 */
-	res = load_elf_segments(utc, elf_state, false /* final attrs */);
-	if (res != TEE_SUCCESS)
-		goto out;
+	for (n = 0; n < num_segs; n++) {
+		res = vm_set_prot(utc, segs[n].va, segs[n].size,
+				  elf_flags_to_mattr(segs[n].flags));
+		if (res)
+			goto out;
+	}
 
 out:
+	free(segs);
 	elf_load_final(elf_state);
 	return res;
```

Comment:
I can't find where `vm_info_final()` gets called when this returns with a error status. Maybe it is not needed (finalized by some other means?). Sorry not being able to point out if it is OK or where is something missing....  :(

Reply:
It's done at end of `ta_load()`.